client/src/stream.rs

```diff
@@ -86,67 +78,30 @@ impl rpccore::Server for CallbackServer {
                 let input_nbytes = nframes as usize * input_frame_size;
                 let output_nbytes = nframes as usize * output_frame_size;
 
-                // Clone values that need to be moved into the cpu pool thread.
-                let mut shm = unsafe { self.shm.as_ref().unwrap().unsafe_view() };
-
-                let duplex_copy_ptr = match &mut self.duplex_input {
-                    Some(buf) => {
-                        assert_eq!(self.dir, StreamDirection::Duplex);
-                        assert!(input_frame_size > 0);
-                        assert!(buf.capacity() >= input_nbytes);
-                        buf.as_mut_ptr()
-                    }
-                    None => ptr::null_mut(),
-                } as usize;
-                let user_ptr = self.user_ptr;
-                let cb = self.data_cb.unwrap();
-                let dir = self.dir;
-
                 // Input and output reuse the same shmem backing.  Unfortunately, cubeb's data_callback isn't
                 // specified in such a way that would require the callee to consume all of the input before
                 // writing to the output (i.e., it is passed as two pointers that aren't expected to alias).
                 // That means we need to copy the input here.
-                let (input_ptr, output_ptr) = match dir {
-                    StreamDirection::Duplex => unsafe {
-                        assert!(input_frame_size > 0);
-                        assert!(output_frame_size > 0);
-                        assert_ne!(duplex_copy_ptr, 0);
-                        let input = shm.get_slice(input_nbytes).unwrap();
-                        ptr::copy_nonoverlapping(
-                            input.as_ptr(),
-                            duplex_copy_ptr as *mut _,
-                            input.len(),
-                        );
-                        (
-                            duplex_copy_ptr as _,
-                            shm.get_mut_slice(output_nbytes).unwrap().as_mut_ptr(),
-                        )
-                    },
-                    StreamDirection::Input => unsafe {
-                        assert!(input_frame_size > 0);
-                        assert_eq!(output_frame_size, 0);
-                        (
-                            shm.get_slice(input_nbytes).unwrap().as_ptr(),
-                            ptr::null_mut(),
-                        )
-                    },
-                    StreamDirection::Output => unsafe {
-                        assert!(output_frame_size > 0);
-                        assert_eq!(input_frame_size, 0);
-                        (
-                            ptr::null(),
-                            shm.get_mut_slice(output_nbytes).unwrap().as_mut_ptr(),
-                        )
-                    },
-                };
+                if let Some(buf) = &mut self.duplex_input {
+                    assert!(input_frame_size > 0);
```

Comment:
```suggestion
                    assert!(input_nbytes > 0);
```
We only care `input_nbytes` after we calculate it from  `input_frame_size`